core/ast_node.js

```diff
@@ -0,0 +1,657 @@
+ * @private
+ */
+Blockly.ASTNode.prototype.findPreviousEditableField_ = function(location,
+    parentInput, opt_last) {
+  var fieldRow = parentInput.fieldRow;
+  var fieldIdx = fieldRow.indexOf(location);
+  var previousField = null;
+  var startIdx = opt_last ? fieldRow.length - 1 : fieldIdx - 1;
+  for (var i = startIdx; i >= 0; i--) {
+    var field = fieldRow[i];
+    if (field.isCurrentlyEditable()) {
+      previousField = field;
+      break;
+    }
+  }
+  return Blockly.ASTNode.createFieldNode(previousField);
+};
+
+/**
+ * Get the first field or connection that is either editable or has connection
+ * value of not null.
+ * @param {!Blockly.Connection} location Current location in the ast.
+ * @param {!Blockly.Input} parentInput The parent input of the field or
+ *     connection.
+ * @return {?Blockly.ASTNode} The ast node holding the next field or connection.
+ * @private
+ */
+Blockly.ASTNode.prototype.findNextForInput_ = function(location, parentInput){
+  var inputs = location.sourceBlock_.inputList;
+  var curIdx = inputs.indexOf(parentInput);
+  if (curIdx <= -1) {return;}
+  var nxtIdx = curIdx + 1;
+  var newAstNode = null;
+
+  for (var i = nxtIdx; i < inputs.length; i++) {
+    var newInput = inputs[i];
+    var field = this.findNextEditableField_(location, newInput, true);
+    if (field) {
+      newAstNode = Blockly.ASTNode.createFieldNode(field);
+      break;
+    } else if (newInput.connection) {
+      var connection = newInput.connection;
+      newAstNode = Blockly.ASTNode.createConnectionNode(connection);
+      break;
+    }
+  }
+  return newAstNode;
+};
+
+/**
+ * Find the next input or field given a field location.
+ * @param {!Blockly.Field} location Current location in the ast.
+ * @param {!Blockly.Input} parentInput The parent input of the field.
+ * @return {?Blockly.ASTNode} The ast node holding the next field or connection.
+ * @private
+ */
+Blockly.ASTNode.prototype.findNextForField_ = function(location, parentInput) {
+  var newAstNode = this.findNextEditableField_(location, parentInput);
+
+  if (!newAstNode || !newAstNode.getLocation()) {
+    newAstNode = Blockly.ASTNode.createConnectionNode(parentInput.connection);
+  }
+  return newAstNode;
+};
+
+/**
+ * Given the current selected field or connection find the previous connection
+ *     or field.
+ * @param {!Blockly.Connection} location Current location in the ast.
+ * @param {!Blockly.Input} parentInput Parent input of the connection or field.
+ * @return {?Blockly.ASTNode} The ast node holding the previous field or
+ *     connection.
+ * @private
+ */
+Blockly.ASTNode.prototype.findPrevForInput_ = function(location, parentInput){
+  var block = location.sourceBlock_;
+  var inputs = block.inputList;
+  var curIdx = inputs.indexOf(parentInput);
+  var newAstNode = null;
+
+  for (var i = curIdx; i >= 0; i--) {
+    var newInput = inputs[i];
+    var fieldNode = this.findPreviousEditableField_(location, newInput, true);
```

Comment:
Minor optimization if you split the two if statements up and move this between them.